### PR TITLE
feat(backend): add workspace scope to memory tool for cross-session persistence

### LINKED ISSUE #162: feat(agent): audit and enhance agent tools — memory, code editing, file operations
## Context

After the LangGraph migration, the agent has 12 tools but several need review and enhancement. The current tool set lacks some capabilities that modern agent systems (Claude Code, OpenClaw, SWE-agent) provide as standard.

### 1. Memory tool (`update_session_config`) — purpose unclear
- Current tool only sets `analysisType`, `designCode`, `skillIds` — essentially session config
- No persistent memory across conversations (no way to remember user preferences, project context, past decisions)
- Need to evaluate: should this be renamed? Should a separate memory/knowledge tool be added?

### 2. No code self-modification capability
- Agent cannot modify its own skill definitions, prompts, or tool implementations
- Claude Code and similar agents can read/edit workspace files including source code
- Consider: should the agent be able to edit Python runtime scripts, skill YAML, or handler.ts files?

### 3. File tools are basic
- `read_workspace_file` / `write_workspace_file` / `list_workspace_files` work but are minimal
- Missing: file search by content (grep), file diff/patch, file rename/move/delete
- Reference implementations: Claude Code has separate Glob (file discovery) + Grep (content search with 3 output modes + pagination)

### 4. No terminal/command execution tool
- OpenClaw and Claude Code can run shell commands in a sandboxed environment
- Useful for: installing Python packages, running validation scripts, debugging analysis failures

## Proposed approach

Audit all 12 existing tools, then design enhancements inspired by:
- **Claude Code**: Glob + Grep + Read + Write + Bash (separate concerns, composable)
- **OpenClaw**: file operations + command execution + code search
- **SWE-agent**: `find_file` + `search_dir` + `search_file` + `edit_file`

### Priority tools to add/fix
1. **Content search tool** (grep in workspace) — high value for debugging
2. **Memory/knowledge tool** — persist user context across conversations
3. **Command execution tool** (sandboxed) — enable package install, script debug
4. **Rename `update_session_config`** — clarify its purpose or split into session config + memory

## Related

- PR #156 (LangGraph migration)
- Current tool definitions: `backend/src/agent-langgraph/tools.ts`

### PR DESCRIPTION
**Pull request by qinsz01**
## Summary

Add a `workspace` memory scope to the agent memory tool alongside the existing `conversation` scope. Workspace entries persist across conversations via a file-backed store (`.runtime/workspace-memory.json`), enabling the agent to retain reusable preferences, durable constraints, and confirmed engineering decisions between sessions.

- **Problem:** Memory was scoped only to individual conversations (`thread_id`). Users had to re-specify preferences (e.g. default design code, project constraints) in every new session.
- **Why it matters:** Workspace-scoped memory enables persistent engineering context that survives session boundaries — a key UX improvement for the chat-first workflow.
- **What changed:** Added `AgentMemoryFileStore` class for JSON file-backed storage; extended `AgentMemoryService` to dispatch by `scopeType` (`conversation` → Prisma/SQLite, `workspace` → file store); added `scope` parameter to the memory tool schema; updated system prompts and tool descriptions for both `en`/`zh`.
- **What did NOT change:** Conversation-scoped memory behavior, Prisma schema, API routes, frontend.

## Change Type (select all)

- [ ] Bug fix
- [x] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [x] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Follow-up to #186 (Conversation-scoped refactor)
- [ ] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

N/A — new feature.

## Regression Test Plan

- Coverage level that should have caught this:
  - [x] Existing coverage already sufficient
- Target tests: `npm test --prefix backend -- --runInBand`
- New tests added:
  - `AgentMemoryFileStore`: store, retrieve, list, delete, file persistence, missing key
  - `AgentMemoryService` workspace scope dispatch
  - `memory-tool` workspace scope: store, retrieve, list, delete, default scope fallback

## User-visible / Behavior Changes

- Memory tool now accepts an optional `scope` parameter: `"conversation"` (default) or `"workspace"`
- System prompts updated to describe both scopes in both `en` and `zh`
- Workspace memory persisted in `.runtime/workspace-memory.json`

## Diagram

```text
Before:  memory tool → conversation scope only (Prisma/SQLite)
After:   memory tool → conversation scope (Prisma/SQLite)
                   ↘ workspace scope  (file store: .runtime/workspace-memory.json)
```

## Security Impact (required)

- New permissions/capabilities? No
- Secrets/tokens handling changed? No
- New/changed network calls? No
- Command/tool execution surface changed? No
- Data access scope changed? No
- If any `Yes`, explain risk + mitigation: N/A

### Steps

1. `npm run db:generate --prefix backend`
2. `npm run build --prefix backend`
3. `npm test --prefix backend -- --runInBand`

### Expected

- All test suites pass
- New `AgentMemoryFileStore` and workspace scope tests pass
- Existing conversation-scoped memory tests unchanged

## Evidence

- [x] New tests added covering the feature

## Human Verification (required)

- Verified scenarios: TypeScript build, Jest full suite including new workspace memory tests
- Edge cases checked: workspace scope without `workspaceRoot` throws, defaults to conversation scope when `scope` omitted, missing key returns null
- What you did **not** verify: Frontend build, E2E tests, LLM integration tests

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? Yes — `conversation` remains the default scope
- Config/env changes? No
- Migration needed? No

## Risks and Mitigations

- Risk: Concurrent writes to `.runtime/workspace-memory.json` could cause data loss
  - Mitigation: Uses atomic write (write to temp file + rename). Single-user dev tool scenario makes concurrent writes unlikely.
- Risk: File store grows unbounded
  - Mitigation: Workspace memory is for preferences/constraints (small, bounded data). Can add size limits later if needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/src/services/agent-memory.ts [hunk 1]
```diff
@@ -1,7 +1,10 @@
+import fs from 'fs';
+import path from 'path';
+import os from 'os';
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts]: Good catch — fixed in 1fcf738. Removed the unused `os` import.

File: backend/src/services/agent-memory.ts [hunk 2]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+  }
+
+  private async readData(): Promise<FileStoreData> {
+    try {
+      const raw = await fs.promises.readFile(this.filePath, 'utf-8');
+      return JSON.parse(raw) as FileStoreData;
+    } catch {
+      return {};
+    }
+  }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts:116]: Valid point — fixed in 1fcf738. `readData()` now only treats `ENOENT` as empty; JSON parse errors and I/O errors are re-thrown to prevent silent data loss.

File: backend/src/services/agent-memory.ts [hunk 3]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+  getFilePath(): string {
+    return this.filePath;
+  }
+
+  private async readData(): Promise<FileStoreData> {
+    try {
+      const raw = await fs.promises.readFile(this.filePath, 'utf-8');
+      return JSON.parse(raw) as FileStoreData;
+    } catch {
+      return {};
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts]: Agreed — same issue raised by Gemini, addressed together in 1fcf738. `readData()` now separates ENOENT from real errors, and surfaces a descriptive error for corrupt JSON.

File: backend/src/services/agent-memory.ts [hunk 4]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+    }
+  }
+
+  private async writeData(data: FileStoreData): Promise<void> {
+    const dir = path.dirname(this.filePath);
+    await fs.promises.mkdir(dir, { recursive: true });
+    const tmp = path.join(dir, `.workspace-memory-${Date.now()}.tmp`);
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts]: Good suggestion — fixed in 1fcf738. Switched to `crypto.randomUUID()` for temp file names.

File: backend/src/services/agent-memory.ts [hunk 5]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+    const data = await this.readData();
+    return Object.entries(data)
+      .map(([key, entry]) => ({
+        scopeType: 'workspace' as const,
+        scopeId: 'default',
+        key,
+        value: entry.value,
+        updatedAt: entry.updatedAt,
+      }))
+      .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts]: Good catch — fixed in 1fcf738. Added `key` as a deterministic tie-breaker so same-millisecond writes always sort consistently by key ascending.

File: backend/src/services/agent-memory.ts [hunk 6]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+    const dir = path.join(workspaceRoot, '.runtime');
+    this.filePath = path.join(dir, 'workspace-memory.json');
+  }
+
+  async store(key: string, value: InputJsonValue): Promise<AgentMemoryEntryView> {
+    const normalizedKey = normalizeMemoryKey(key);
+    const data = await this.readData();
+    const updatedAt = new Date().toISOString();
+    data[normalizedKey] = { value: value as JsonValue, updatedAt };
+    await this.writeData(data);
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts:46]: Same reasoning as the Gemini comment above — the agent runtime is single-threaded per conversation. Two concurrent `store()` calls on the same file store instance cannot happen in the current architecture. Adding serialization here would be speculative complexity for a scenario the system cannot produce today.

File: backend/src/services/agent-memory.ts [hunk 7]
```diff
@@ -16,8 +19,118 @@ export interface AgentMemoryEntryView {
...
+    return Object.entries(data)
+      .map(([key, entry]) => ({
+        scopeType: 'workspace' as const,
+        scopeId: 'default',
+        key,
+        value: entry.value,
+        updatedAt: entry.updatedAt,
+      }))
+      .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
+  }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts:85]: Not adding a hard cap here. Workspace memory is for small bounded data (preferences, constraints, design codes) — typically a handful of entries. The conversation-scoped `list` uses Prisma's `take: 50` only because it queries a shared SQLite table that could accumulate across many conversations. The file-backed workspace store is scoped to a single workspace instance and will not grow unbounded in practice. If this becomes a concern, a cap can be added later without API changes.

File: backend/src/services/__tests__/agent-memory.test.mjs [hunk 8]
```diff
@@ -35,3 +38,100 @@ describe("AgentMemoryService", () => {
...
+
+  test("lists entries sorted by updatedAt desc", async () => {
+    await fileStore.store("alpha", { v: 1 });
+    await fileStore.store("beta", { v: 2 });
+    const entries = await fileStore.list();
+    expect(entries.length).toBeGreaterThanOrEqual(2);
+    // newest first
+    expect(entries[0].key).toBe("beta");
+    expect(entries[1].key).toBe("alpha");
+  });
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/__tests__/agent-memory.test.mjs:82]: Fixed in 1fcf738 — the `list()` sort now uses `key` as a deterministic tie-breaker, so the test ordering is stable regardless of timestamp collision.

File: backend/src/services/agent-memory.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/services/agent-memory.ts:46]: Acknowledged, but deliberately not adding a lock here. This is a single-user local dev tool — the LangGraph ReAct agent processes one request at a time. There is no realistic concurrency scenario where two `store()` calls race on the same `AgentMemoryFileStore` instance. Adding a mutex would add complexity for a problem that doesn't exist in the actual runtime. If this ever becomes multi-user, a lock would be warranted at that point.

**Review by guyi2000 — APPROVED** [review]
